homeassistant/components/lock/bmw_connected_drive.py

```diff
@@ -0,0 +1,119 @@
+"""
+Support for BMW cars with BMW ConnectedDrive.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/lock.bmw_connected_drive/
+"""
+import asyncio
+import logging
+import time
+
+from homeassistant.components.bmw_connected_drive import DOMAIN as BMW_DOMAIN
+from homeassistant.components.lock import LockDevice
+from homeassistant.const import STATE_LOCKED, STATE_UNLOCKED
+
+DEPENDENCIES = ['bmw_connected_drive']
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the BMW Connected Drive lock."""
+    accounts = hass.data[BMW_DOMAIN]
+    _LOGGER.debug('Found BMW accounts: %s',
+                  ', '.join([a.name for a in accounts]))
+    devices = []
+    for account in accounts:
+        for vehicle in account.account.vehicles:
+            device = BMWLock(account, vehicle, 'lock', 'BMW lock')
+            devices.append(device)
+    add_devices(devices)
+
+
+class BMWLock(LockDevice):
+    """Representation of a BMW vehicle lock."""
+
+    def __init__(self, account, vehicle, attribute: str, sensor_name):
+        """Initialize the lock."""
+        self._account = account
+        self._vehicle = vehicle
+        self._attribute = attribute
+        self._name = sensor_name
+        self._state = None
+
+    @property
+    def should_poll(self):
+        """Do not poll this class.
+
+        Updates are triggered from BMWConnectedDriveAccount.
+        """
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the lock."""
+        return self._name
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the lock."""
+        vehicle_state = self._vehicle.state
+        return {
+            'last_update': vehicle_state.timestamp,
+            'car': self._vehicle.modelName,
+            'door_lock_state': vehicle_state.door_lock_state.value
+        }
+
+    @property
+    def is_locked(self):
+        """Return true if lock is locked."""
+        return self._state == STATE_LOCKED
+
+    def lock(self, **kwargs):
+        """Lock the car."""
+        from bimmer_connected.remote_services import ExecutionState
+
+        _LOGGER.debug("%s: locking doors", self._vehicle.modelName)
+        status = self._vehicle.remote_services.trigger_remote_door_lock()
+        self._state = STATE_LOCKED
```

Comment:
Either the update callback should handle setting state or we should do an optimistic state update from here, immediately. The logic below doesn't belong in this method.